### ISSUE TITLE
using constant method

### ISSUE THREAD

**Issue by manuelmolano** [nicholas-roy/psytrack#12]
https://github.com/nicholas-roy/psytrack/blob/d0ee66f16e6735f30145b5f8a7bd2e671f3cd9ce/psytrack/getMAP.py#L241

Hello,

I'd like to use the 'constant' method to prevent the sigmas from getting very large values. Would this be possible?

Thanks for the help!

**Comment by nicholas-roy** [comment]
Hi,

Sorry, most of the code surrounding the `method` parameter is a vestige of an earlier version of Psytrack (which I probably should go back and fix...). However, the "constant" method is (almost) equivalent to running vanilla logistic regression on all the data at once (so you would get K weights, each with a single value that applies to all trials). Another way to think of it is that you would be forcing your weights to be perfectly flat. 

I believe SciPy has an implementation of logistic regression if that is what you are looking to do. If you are not looking to do that, could you explain your situation in a bit more detail?  I'm not really sure what you mean when you say that your sigma are too large?

Best,
Nick

**Comment by manuelmolano** [comment]
Hi Nick,

Thanks for your fast response. I think I misunderstood what the constant method does. I just want to fix (or bound) the sigmas so they do not grow too large. 
Looking at [the psytrack tutorial](https://github.com/nicholas-roy/psytrack/blob/master/psytrack/examples/ExampleNotebook.ipynb), it seems one can get this just by passing an empty optList to the hyperOpt function. However when I do that I get the following error:

**ValueError: zero-size array to reduction operation maximum which has no identity**

Looking at the code it seems the issue is in the fact that, being optList empty, optVals is empty and the scipy minimize function does not like it. Am I doing something wrong?
thanks again!

**Comment by nicholas-roy** [comment]
Thanks for the added info! If passing an empty `optList` is not working, than that's a bug! 

To be clear, what that _should_ do is optimize the weights without optimizing the sigmas, i.e. it assumes that the sigmas provided are fixed. You can also achieve this by calling the `getMAP` function directly (which is what I suggest you do to get the functionality you want, while I fix the bug above).

However, you seem to be indicating that you have tried to optimize the sigmas and they get too big (too negative?). I can't say I've ever seen this happen before, so I suspect there is something else wrong, if this is the case.

**Comment by manuelmolano** [comment]
I see, I will try directly using the getMAP function then.
Regarding the sigmas, it's not that they get unreasonably big (they are not negative), but when applying the tool to different subjects I get much variability in the magnitude of the sigmas from subject to subject and so I was wondering whether it was possible to limit somehow that variability to get more comparable results. Does this make sense?

**Comment by nicholas-roy** [comment]
If you're confident that the variability in the sigmas between patients is not actually a feature of the data (i.e. I've seen lots of variation in sigmas within populations of rats and mice), then I agree that fixing sigma make sense!

**Comment by manuelmolano** [comment]
Yes, you're right. I just wanted to check whether the main results changed a lot if I fixed the sigmas.

Thanks for the help!

**Comment by jpillow** [comment]
Thanks for your message, Manuel.

I can certainly imagine that if the # of trials per subject is not *that*
high, then there could be a lot of variability in the sigmas.  It would
actually be nice if the use could specify a prior on the sigmas, e.g., a
Gaussian prior on log(sigma), which would keep them from varying too much.

Nick, is this possible under the current implementation?  I imagine you
could just have the user pass in a function they want to use as a sigma
prior, which would give them the flexibility to decide for themselves.
(e.g. if you passed in a uniform prior it would amount to a range
constraint on the sigmas).

best,
Jonathan

**Comment by nicholas-roy** [comment]
Well perhaps the easiest thing to do would be to pass explicit bounds to the optimizer itself. 

Specifically, on [this line](https://github.com/nicholas-roy/psytrack/blob/master/psytrack/hyperOpt.py#L181)

        result = minimize(
            hyperOpt_lossfun,
            optVals,
            args=opt_keywords,
            method='BFGS',
            options=opts,
            callback=callback,
        )

you can change the optimization method from 'BFGS' to 'L-BFGS-B', which then allows you to pass a `bounds` argument to the optimizer for each hyperparameter (see [`scipy.optimize.minimize`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html)).

**Comment by manuelmolano** [comment]
Hi, thanks for the suggestion. I made some quick tests and it seems it works but for some reason the final sigmas are not different from the initial ones. I will need to understand the L-BFGS-B method a bit better, but I just mention this issue in case you guys finally decide to modify the psy-track so it allows bounding somehow the sigmas. 
I think this would be a useful functionality, since it would avoid the dilemma between using the standard GLM (i.e. sigmas=0) and using psy-track (sigmas potentially big, and noisy traces) by allowing a smooth transition between the two methods.
Hope this helps!
thanks!
Manuel
